### PR TITLE
8290376: G1: Refactor G1MMUTracker::when_sec

### PR DESCRIPTION
**Pull request by albertnetymk**
Introducing a more intuitive implementation.

This PR consists of two commits: the first adds the new algorithm and verifies the result is same as before. The second commit removes the old implementation.

Test: tier1-6 for the first commit.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290376](https://bugs.openjdk.org/browse/JDK-8290376): G1: Refactor G1MMUTracker::when_sec

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [a52a6f54](https://git.openjdk.org/jdk/pull/9517/files/a52a6f548f51e1c73a6bb507f537082f17cd9402)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [a52a6f54](https://git.openjdk.org/jdk/pull/9517/files/a52a6f548f51e1c73a6bb507f537082f17cd9402)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9517/head:pull/9517` \
`$ git checkout pull/9517`

Update a local copy of the PR: \
`$ git checkout pull/9517` \
`$ git pull https://git.openjdk.org/jdk pull/9517/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9517`

View PR using the GUI difftool: \
`$ git pr show -t 9517`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9517.diff">https://git.openjdk.org/jdk/pull/9517.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 1]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
 double G1MMUTracker::when_sec(double current_timestamp, double pause_time) {
-  // If the pause is over the maximum, just assume that it's the maximum.
-  double adjusted_pause_time =
-    (pause_time > max_gc_time()) ? max_gc_time() : pause_time;
+  assert(pause_time > 0.0, "precondition");
 
-  // Earliest end time of a hypothetical pause starting now, taking pause_time.
-  double earliest_end_time = current_timestamp + adjusted_pause_time;
-  double gc_time_in_recent_time_slice = calculate_gc_time(earliest_end_time) + adjusted_pause_time;
+  // Clamp it by max
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:136]:
```suggestion
    // If the pause is over the maximum, just assume that it's the maximum.
```
I would prefer the original comment as it explicitly states that this is a conscious decision.

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 2]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
+//              |<- limit |                   |
+//              |         |<- deficit         |
+//              |         ^                   |
+//              |                             |
+//              |<------- _time_slice   ----->|
+//
+// Focusing on GC events that are inside [limit, current_timestamp], we iterate
+// over them from the newest to the oldest (right-to-left in the diagram) and
+// try to locate the timestamp annotated with ^, so that the accumulated GC
+// time inside [deficit, current_timestamp], is equal to gc_budget. Next,
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:129]: `gc_budget` is not defined anywhere in this comment or the diagram. This is confusing.

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 3]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
-    G1MMUTrackerElem *elem = &_array[_head_index];
-    return (elem->end_time() + (_time_slice - max_gc_time())) - current_timestamp;
-  }
-
-  // Now go through the recent pause time events,
-  double limit = earliest_end_time - _time_slice;
-  int index = _tail_index;
-  while ( 1 ) {
+  double limit = current_timestamp + pause_time - _time_slice;
+  // Iterate from newest to oldest
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:142]:
```suggestion
  // Iterate from newest to oldest.
```

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 4]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
-    if (elem->end_time() > limit) {
-      if (elem->start_time() > limit) {
-        gc_time_to_pass -= elem->duration();
-      } else {
-        gc_time_to_pass -= elem->end_time() - limit;
-      }
-      if (is_double_leq_0(gc_time_to_pass)) {
-        return elem->end_time() + (_time_slice + gc_time_to_pass) - earliest_end_time;
-      }
+    // Outside the window
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:146]:
```suggestion
    // Outside the window.
```

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 5]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
+    if (elem->end_time() <= limit) {
+      break;
+    }
+
+    double duration = (elem->end_time() - MAX2(elem->start_time(), limit));
+    // Would exceed the budget; strictly greater than
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:152]:
```suggestion
    // This duration would exceed the budget. Check for strictly greater than.
```
Not sure what the "[Check for ]strictly greater than" adds, but it's okay.

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 6]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
+    if (duration > gc_budget) {
+      // The timestamp where a budget deficit occurs.
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:154]:
```suggestion
      // This GC event contains the timestamp where the budget deficit occurs.
```

File: src/hotspot/share/gc/g1/g1MMUTracker.cpp [hunk 7]
```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
...
+      assert(deficit_timestamp >= limit, "inv");
+      return deficit_timestamp - limit;
     }
-    index = trim_index(index+1);
-    guarantee(index != trim_index(_head_index + 1), "should not go past head");
+
+    gc_budget -= duration;
   }
+
+  // No enough gc time inside the window; a budget surplus
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1MMUTracker.cpp:163]:
```suggestion
  // Not enough gc time spent inside the window, we have a budget surplus.
```

**Review by tschatzl — CHANGES_REQUESTED** [review]

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Review by walulyai — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**